### DISCUSSION TITLE
Updating new listeners clients (OTA)

### DISCUSSION THREAD

**Discussion by MatteoGheza** [josephdadams/TallyArbiter#267] [General]
Now listener clients are in this repo and works with the new listenerclient_connect api (see #225 and leave a review, please).
Now we should detect an old client (I'm working on this, listening to socket events used by old clients and displaying a banner in devices page) and require a client update.
Client update isn't an easy process, especially if you are using clients in production.
We should use OTA updates, so this is an one-time process.
How can we release a new listener version and update the device?
@hrueger any thoughts?

**Comment by hrueger** [comment]
Hm, difficult question. I don't know about the embedded side.
I've been doing some updating of STM32 microcontrollers lately, but that was over usb and their DFU mode. I don't know how we could do that over wifi.

I can, however, say that I would prefer to ship the firmware bin files with TallyArbiter. Then, updating also works offline.

**Comment by josephdadams** [comment]
@mattv8 any thoughts on OTA here? I was thinking you were working on something with this.

**Reply by mattv8** [reply]
Hey @josephdadams, yep, I'm currently working on developing OTA for my listener client but my time is oh so limited these days so I can't give an accurate ETA on that. I'm going to base my OTA strategy off of a really good write-up on [RandomNerdTutorial](https://randomnerdtutorials.com/esp32-ota-over-the-air-arduino/) using the [AsyncElegantOTA library](https://github.com/ayushsharma82/AsyncElegantOTA). The question then is how difficult would it be to hook TallyArbiter into the AsyncElegantOTA web UI? I can't imagine it would be too difficult!

I'll plan on contributing to the OTA updating portion of the listener client code as I have availability. I'll update as I figure stuff out, but if anyone else wants to look into it in parallel that would sure speed things up.
